eth/db/header.py

```diff
@@ -218,10 +225,103 @@ def _persist_checkpoint_header(
             header.hash,
             rlp.encode(header),
         )
+
+        # Add new checkpoint header
+        checkpoint_headers_key = SchemaV1.make_checkpoint_headers_key()
+        previous_checkpoints = db.get(checkpoint_headers_key) or b''
+        db.set(
+            checkpoint_headers_key,
+            previous_checkpoints + header.hash,
+        )
+
         previous_score = score - header.difficulty
         cls._set_hash_scores_to_db(db, header, previous_score)
-        cls._set_as_canonical_chain_head(db, header, header.parent_hash)
-        cls._update_header_chain_gaps(db, header)
+        cls._set_as_canonical_chain_head(db, header, GENESIS_PARENT_HASH)
+        _, gaps = cls._update_header_chain_gaps(db, header)
+
+        # check if the parent block number exists, and is not a match for checkpoint.parent_hash
+        parent_block_num = BlockNumber(header.block_number - 1)
+        try:
+            parent_hash = cls._get_canonical_block_hash(db, parent_block_num)
+        except HeaderNotFound:
+            # no parent to check
+            pass
+        else:
+            # User is asserting that the checkpoint must be canonical, so if the parent doesn't
+            # match, then the parent must not be canonical, and should be de-canonicalized.
+            if parent_hash != header.parent_hash:
+                # does the correct header exist in the database?
+                try:
+                    true_parent = cls._get_block_header_by_hash(db, header.parent_hash)
+                except HeaderNotFound:
+                    # True parent unavailable, just delete the now non-canonical one
+                    cls._decanonicalize_into_gap(db, parent_block_num, gaps)
+                else:
+                    # True parent should have already been canonicalized during
+                    #   _set_as_canonical_chain_head()
+                    raise ValidationError(
+                        f"Why was a non-matching parent header {parent_hash} left as canonical "
+                        f"after _set_as_canonical_chain_head() and {true_parent} is available?"
+                    )
+
+        cls._decanonicalize_naughty_children(db, header)
+
+    @classmethod
+    def _decanonicalize_naughty_children(cls, db: DatabaseAPI, header: BlockHeaderAPI) -> None:
```

Comment:
I can't remember if it's just my own terminology but I've used the term *orphaned* for this previously (to refer to headers that were one part of the chain but are no longer canonical).

Reply:
Heh yeah, maybe `_decanonicalize_new_orphans()` is a better name.